### DISCUSSION TITLE
Patch data resolver for Feathers Schema

### DISCUSSION THREAD

**Discussion by RuneSP** [feathersjs/feathers#2658] [General]
Hi, 

I'm tempted to switch from mongoose models to use feathers schemas instead, as its much easier to share interfaces, validation etc with frontend.  

I'm currently stuck on how to correctly create a patch resolver, the examples only mentions an "userPatchResolver". 

If we take the UserSchema from the [examples page](https://dove.docs.feathersjs.com/api/schema/resolvers.html#data-resolvers). If I use the userDataResolver for a patch event and only updates the user email, it will throw an error with password missing, as it should.

I guess I could make a userPatchSchema, that extends userSchema and remove password from required properties, but this will still throw an error if not all other required properties are in the patch data. 

I think the patch resolver should only resolve / validate the properties that are present in the patch data, but I'm not really sure if thats possible?

Thanks

**Comment by daffl** [comment]
The new CLI [uses extension](https://dove.docs.feathersjs.com/api/schema/schema.html#extension) setting the `required` property to an empty array which means fields that do not exist don't get validated:

```ts
import { schema } from '@feathersjs/schema';

export const userDataSchema = schema({
  $id: 'UserData',
  type: 'object',
  additionalProperties: false,
  required: ['email', 'password'],
  properties: {
    id: { type: 'number' },
    email: { type: 'string' },
    password: { type: 'string' }
  }
});

export const userPatchSchema = schema({
  $id: 'UserPatch',
  type: 'object',
  additionalProperties: false,
  required: [],
  properties: {
    ...userDataSchema.properties
  }
});
```

**Comment by RuneSP** [comment]
That works, but will still accept null values (and convert them to empty string, for strings). I guess I can set minLength to prevent this for strings.

Another question, feathers / json schema doesnt seem to handle uniqueness of properties. I can achieve this with hooks, but would rather have it set at database level. What would be the feathery way to achieve this using schemas?

**Reply by daffl** [reply]
Interesting, I didn't know it did that with null values. I wonder if there is an AJV setting to prevent that.

Uniqueness should be handled on the database level whenever possible but can also be done in Feathers with a hook that makes a `find` call for the field that should be unique and checks the `total`:

```ts
import { Conflict } from '@feathersjs/errors'

app.service('myservice').hooks({
  before: {
    create: [async context => {
      const { name } = context.data

      const { total } = await context.service.find({
        query: {
          // This means it will just do a quick count to return the total
          $limit: 0,
          name
        }
      })

      if (total > 0) {
        throw new Conflict(`Entry with ${name} already exists`)
      }
    }]
  }
})
```
